### ISSUE TITLE
Database model

### ISSUE THREAD

**Issue by dotNETSanta** [CoyoteNET/CoyoteNET-backend#31]
So since I started implementing our db model I'd want to talk about a few things:

User lacks of "Email"

Why does post have an "Title"? I'd say that "Thread" should've it.

File shouldnt be held in Database, thus we gotta change it into something like:

	{
		Id = 60f67d69-5b84-4ded-81b2-45c668d21c85
		DiskFileName = f3d9292a-8179-43dc-b72c-4bb3be6e78ff
		UserFileName = "123.jpg",
		UploadedBy = ...
		DownloadedBy[] ...
	}
	
Email lacks of EmailType

Email Sender - hmm? that's weird. What's even the purpose of that? are poeple going to be able to send emails from our our forum? email sender is always system (I suppose)

I'd change all those ``Notification*`` entities into one and add ``NotificationType``

User shouldn't store password, but instead we should store ``PasswordHash`` and ``PasswordSalt`` (afaik)

What the purpose of ``ForumSection`` when we already have ``ForumCategory``?

**Comment by silvuss** [comment]
> So, since I started implementing our db model I'd want to talk about a few things:
> 
>     * User lacks of "Email"

That is true. My miss. I will add it.
 
>     * Why does post have an "Title"? I'd say that "Thread" should've it.

That is true. My miss. I will add it.

>     * File shouldnt be held in Database, thus we gotta change it into something like:
> 
> > 
> 
> ```
> {
> 	Id = 60f67d69-5b84-4ded-81b2-45c668d21c85
> 	DiskFileName = f3d9292a-8179-43dc-b72c-4bb3be6e78ff
> 	UserFileName = "123.jpg",
> 	UploadedBy = ... // that's just suggestion, but I'm not sure whether we'll even need it
> 	DownloadedBy[] ... // that's just suggestion, but I'm not sure whether we'll even need it
> }
> ```

Files **could** be held in the database but if you are not about it, I agree. That is just an option.

>     * `E-mail messages sending log` lacks of EmailType

That is true. My miss. I will add it.

>     * `E-mail messages sending log` **Email Sender** - hmm? that's weird. What's even the purpose of that? are poeple going to be able to send emails from our our forum? email sender is always system (I suppose)

E-mail may be send only by privileged users (moderators and administrators). The attribute "sender" is to track who send an e-mail, which privileged user has "ordered" it.

>     * I'd change all those `Notification*` entities into one and add `NotificationType`

That is an option, right. But I thought that it would be OK to split the entities because I suppose that there will be many notifications (it will be half the usual database traffic, I suppose), so to increase efficiency, multiple tables might be used asynchronously.

>     * User shouldnt have password, but instead we should store `PasswordHash` and `PasswordSalt` (afaik)

That may be true, I do not know.

>     * What the purpose of `ForumSection` when we already have `ForumCategory`?

A "forum section" is something like "Programowanie - języki i technologie" in the original Coyote system. Assumedly, creation, deletion and name edition of such "sections" shall be less "tracked" in the database than forum categories – that is, there shall be more freedom in it. Its purpose is to better organize forum structure, because, as in the original Coyote, there might someday be more than few categories.

**Comment by dotNETSanta** [comment]
``E-mail may be send only by privileged users (moderators and administrators). The attribute "sender" is to track who send an e-mail, which privileged user has "ordered" it.``

Do we actually need it? does it exist on 4p?

**Comment by silvuss** [comment]
@dotNETSanta I do not know whethere the original Coyote has such an option (I suppose not). Maybe somekind could answer, but unfortunately he is not in this project (I cannot mention him). That may be good enhancement, how do you think?

**Comment by dotNETSanta** [comment]
@silvuss 

Personally I can't think of any reason why moderator would rather send email to somebody via forum.

On the other hand it would be viable option for administrator, but why would we want to write web email client? oO it isn't as easy as it seems to be.

**Comment by dotNETSanta** [comment]
What's the purpose of "Thread category notifications"?

**Comment by silvuss** [comment]
@dotNETSanta That may be true that we would not want to write a separate e-mail client... ;) Maybe there are options to integrate us with an existing one. I understand "enhancement" in a way that one may easily find out who send what e-mail to what user and when.

"Thread category notifications" are to track a thread category for each new thread. I remember that such an option was mentioned in the comments on 4p at least once or twice.

**Comment by dotNETSanta** [comment]
@silvuss  User using "@silvuss"

>But I thought that it would be OK to split the entities because I suppose that there will be many notifications (it will be half the usual database traffic, I suppose), so to increase efficiency, multiple tables might be used asynchronously.

So, you're suggesting that we should create let's say 4 tables where each of them holds different notifications

and we're constantly searching in 4 tables in order to find Notification for given user? I think that it'll hurt performance, but we gotta ask experts hehe @smknd

**Comment by silvuss** [comment]
> @silvuss User using "@silvuss"
> 
> > But I thought that it would be OK to split the entities because I suppose that there will be many notifications (it will be half the usual database traffic, I suppose), so to increase efficiency, multiple tables might be used asynchronously.
> 
> So, you're suggesting that we should create let's say 4 tables where each of them holds different notifications
> 
> and we're constantly searching in 4 tables in order to find Notifications for given user? I think that it'll hurt performance, but we gotta ask experts hehe @smknd

Yes, now I see, updating tables will have better performance, but sending the notifications to users will not. :P I do not know for now how to solve it, I will think about it.

Please, send a pull request with all your changes to data models page at the wiki because I would like it to be exactly the same as the implementation. I will review it, and we may discuss further changes there.

**Comment by silvuss** [comment]
@smknd, @dotNETSanta, I have created the issue #36 to talk about the benefits of having 4 tables instead of 1.

**Comment by silvuss** [comment]
@dotNETSanta ,
> Are our **Editions [Entity]** doing actually what we want? For me they're now just being another "Logger" which loggs that some user commited an edit on some Entity but is it useful for us somehow?
> 
> I see two approaches to that:
> 
> Entity holds who commited last change and when
> 
> or
> 
> (seems like we're doing this one atm) We actually have an history of everything who, when and what was changed of that entity. We currently we do not log "what was changed".

that is a good point. Indeed, currently relations "Editions of ..." (please, use the word "relation" – an "entity" is something different) are not designed for tracking **what** was changed.

I thought that such tracking makes no sense. For example, a user changes one letter their post, and after a minute, they change two letter, and after a two minutes, they change one letter again... But I may be wrong. @dotNETSanta do you have some idea how tracking the "what" could look like?

**Comment by dotNETSanta** [comment]
>I thought that such tracking makes no sense. For example, a user changes one letter in their post, and after a minute, they change two letter, and after a two minutes, they change one letter again... But I may be wrong that it is just that simple.

>PS2. I forgot to write what is the purpose of having such tables since that is also your concern. When creating the model, I thought that it would be useful for us – in general. I did not think about any particular use cases. Thinking now, we can be displaying information about editions at all: "Edited"; or, information similar to that in original Coyote or Stack Overflow: "Edited [date] by [user], edited [date2] by [admin/user] etc."

I'm not really convinced of the usefulness of that.

btw: github has history of e.g comment edition under [edited] button. check that

**Comment by silvuss** [comment]
Yes, I have seen the GitHub history. GitHub could handle such big amount of data, bu could we?

**Comment by dotNETSanta** [comment]
@silvuss Nothing's preventing us from saving only last X edits.

**Comment by silvuss** [comment]
@dotNETSanta That is true, I did not think about it before. If there would be any real benefits for us or users of the system of doing so, maybe we could. Maybe 5 to 10.

**Comment by KrzysztofUrbanczyk** [comment]
`Threads`, `Posts`, and `MicroblogEntry` should contain a list of subscribers.
Maybe the better name for `Thread` is `Topic`?

**Comment by dotNETSanta** [comment]
>Maybe the better name for Thread is Topic?

Isn't ``Topic`` just a name of a thread? e.g "How to use loops in C#?"

>Threads, Posts, and MicroblogEntry should contain a list of subscribers.

That's great remark @KrzysztofUrbanczyk 

I tried to add those while using ``Discriminator`` approach instead of  using something like "SubscriptionType" to determine what kind of subscription it is (thread, post or microblog) and I'd really want to hear your opinions guys (especially @smknd) on that

Here's commit: https://github.com/CoyoteNET/CoyoteNET-backend/commit/1b6d3ee35997cc75b9bfc0b4f5e5a7d090bf1673

And it'll be working only with ``InMemoryDb`` because I thought that configuring every entity manually is easier, haha. I just tried to create that db on MSSQL and apparently I'mma to spend a lot of time with that

**Comment by KrzysztofUrbanczyk** [comment]
OK, `Thread` is better.

I also play in the project and just add:

`public List<User> Subscribers { get; } = new List<User>(); `

I see `Discriminator` for the first time but looks good for me.

Could you not use shortcuts? e.g:
use `user` instead of `u`
`public SubscriptionPost(User u) : base(u)`

**Comment by dotNETSanta** [comment]
@KrzysztofUrbanczyk 

>I also play in the project and just add:

`public List<User> Subscribers { get; } = new List<User>(); `

wait, what?

> { get; } 

If I recall correctly, then EF require you to have an **setter**.

>Could you not use shortcuts? e.g:
>use user instead of u
>public SubscriptionPost(User u) : base(u)

Yea, sure.

**Comment by dotNETSanta** [comment]
Now it at least manages to create our db with our not finished-yet model.

Some things that caused troubles I just commented out and it'd be nice if somebody could look at those relations between ``User`` and ``UserFile`` and ``SubscriptionThread``

Here's code map and db diagram

https://i.imgur.com/A9gLLu2.png

https://i.imgur.com/A9gLLu2.png

**Comment by dotNETSanta** [comment]
@silvuss 

What do you think about ``BanExpirationDate`` which's nullable property of User

When user receives ban then this value is being set to some date that'll represent duration of ban.

**Comment by dotNETSanta** [comment]
>6 [silvuss's note] Originally, there was a relation "Threads' tags" that worked like an associative relation between the relations "Threads" and "Tags". That made the whole model "more compatible" with the 1st normal form (i.e., all values are atomic). But I have realized that all threads may have tags associated, and there could be many threads with at least a couple of tags. Since searching such a table shall be the basic operation, I have concluded that it would be harmful for database efficiency. As a result, it makes existing of the whole relation pointless at all. Instead, now the attribute Tags contains a character string. It MUST be a concatenation of several character strings, each representing a tag. The concatenation MUST be performed by commas. Each tag MUST be included in the relation "Tags" as a separate tuple.

Are wer sure about that?

How we will do things like "find all threads (with pagination) for tag "java""?

we will split that string by "," and check whether any element of that split is "java"? I think just many 2 many relation and performing **join** would be faster and it'd be easier way to maintain.

**Comment by dotNETSanta** [comment]
Can we rename ``Thread`` to ``ForumThread``? There's already existing https://docs.microsoft.com/en-us/dotnet/api/system.threading.thread?view=netframework-4.8 and we're also using it heavily, so I often need to clarify which one I'm using

**Comment by KrzysztofUrbanczyk** [comment]
That's why I wanted to change the name to `Topic`.
